### ISSUE TITLE
NullReference on iOS Xamarin

### ISSUE THREAD

**Issue by NVentimiglia** [ably/ably-dotnet#112]
Here is our caught stack trace.
```

System.NullReferenceExceptionObject reference not set to an instance of an object
Raw
WebSocket4Net.WebSocket.ClearTimer()
WebSocket4Net.WebSocket.FireClosed()
WebSocket4Net.WebSocket.OnClosed()
WebSocket4Net.WebSocket.client_Error(object sender, ErrorEventArgs e)
SuperSocket.ClientEngine.ClientSession.OnError(Exception e)
SuperSocket.ClientEngine.SslStreamTcpSession.OnDataRead(IAsyncResult result)
System.Net.LazyAsyncResult.Complete(IntPtr userToken)
System.Net.LazyAsyncResult.ProtectedInvokeCallback(object result, IntPtr userToken)
System.Net.LazyAsyncResult.InvokeCallback(object result)
Mono.Net.Security.AsyncProtocolRequest.StartOperation()
Mono.Net.Security.AsyncProtocolRequest.<StartOperation>b__17_0(object _)
System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(object state)
System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, object state, bool preserveSyncCtx)
System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, object state, bool preserveSyncCtx)
System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
System.Threading.ThreadPoolWorkQueue.Dispatch()
ObjCRuntime.Runtime.ThreadPoolDispatcher(Func<TResult> callback)
System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Comment by mattheworiordan** [comment]
@NVentimiglia can you confirm which version this is?
Also, have you tried the latest version that no longer relies on WebSocket4Net and instead uses the MS WebSocket implementation?  Please can you try that and confirm if the issue still exists.

**Comment by marto83** [comment]
@NVentimiglia I have a suspicion that this is caused by the phone losing connectivity. I'm going to look into finding a way to log and handle connectivity changes instead of letting the connection error. 

Still, please give the latest beta a try as the protocol has been completely changed.

**Comment by NVentimiglia** [comment]
@marto83 
8.5.6

**Comment by caseyniemann** [comment]
ConnectAsyncExtension.SocketConnectCompleted(System.Object sender, System.Net.Sockets.SocketAsyncEventArgs e)
android.runtime.JavaProxyThrowable: System.NullReferenceException: Object reference not set to an instance of an object

Xamarin caused by: android.runtime.JavaProxyThrowable: System.NullReferenceException: Object reference not set to an instance of an object
SuperSocket.ClientEngine.ConnectAsyncExtension.SocketConnectCompleted(object sender, SocketAsyncEventArgs e)<2b3384da5fba458d95aaf3e5bd2b26a7>:0
SuperSocket.ClientEngine.ConnectAsyncExtension.OnGetHostAddresses(IAsyncResult result)<2b3384da5fba458d95aaf3e5bd2b26a7>:0
System.Runtime.Remoting.Messaging.AsyncResult.Invoke(AsyncResult)(wrapper managed-to-native)
System.Runtime.Remoting.Messaging.AsyncResult.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()<b372b86075aa499fb19b0e7f203d3426>:0
System.Threading.ThreadPoolWorkQueue.Dispatch()<b372b86075aa499fb19b0e7f203d3426>:0
System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()<b372b86075aa499fb19b0e7f203d3426>:0

**Comment by caseyniemann** [comment]
Above is crash report from 8.4.6

**Comment by caseyniemann** [comment]
updated to 8.4.7 in our Dev---still getting a lack o signals when we switch networks from LTE to WIFI and back again...it's as though the signal retry (10 seconds) in our case, doesn't immediately recognize if the network connection has changed from LTE to WIFI

**Comment by mattheworiordan** [comment]
FYI, just chatted with Martin @caseyniemann. We would like to work with the latest version of ably-dotnet to get this working and not use an older version as there is little point in supporting an old version.  @marto83 will be in touch soon if he has any questions or needs you to test anything.  Tanks for bearing with us

**Comment by marto83** [comment]
@caseyniemann we are moving away from WebSocket4Net. Can you try with the latest beta? I'm looking into handling Mobile OS network service events but it definitely won't be in 0.8.4.*.

**Comment by NVentimiglia** [comment]
0.8.5-beta8 did not send / receive on IOS

**Comment by mattheworiordan** [comment]
@NVentimiglia what do you mean by `0.8.5-beta8 did not send / receive on IOS?  What specifically happened, and is this related to the reported issue of changing networks?

**Comment by caseyniemann** [comment]
@marto83 @mattheworiordan the beta didn't signal at all on IOS...it did on android.

**Comment by caseyniemann** [comment]
@marto83 can we connect live and debug

**Comment by marto83** [comment]
@caseyniemann I've sent you an email. Can you enable debug logging and post the logs here from iOS.

**Comment by ndastur** [comment]
Just to note that I am getting the same error as Crash Reports from Hockey. So haven't yet seen in local debug. Using 0.8.4.7

**Comment by marto83** [comment]
This is fixed in the latest beta when using the WebSocket4Net transport. 
The library will handle the error and continue to shut down the websocket transport after which it will try and create a new connection.

**Comment by marto83** [comment]
This is fixed in 0.8.5 beta 18. Feel free to reopen it if it reoccurs.
